File: scripts/Data/Action.cs
using System;
using System.Collections.Generic;
using Godot;

namespace DND;

public partial class Action : Node
{
    public String Description { get; set; }
    public Stat required_stat
    {
        get => RequiredStat;
        set => RequiredStat = value;
    }
    public Stat RequiredStat { get; set; }

    public Outcome positive_outcome
    {
        get => PositiveOutcome;
        set => PositiveOutcome = value;
    }

    public Outcome negative_outcome {
        get => NegativeOutcome;
        set => NegativeOutcome = value;
    }
    public Outcome PositiveOutcome { get; set; }
    public Outcome NegativeOutcome { get; set; }

    public override string ToString()
    {
        return $"Description: {Description}\n\r" +
               $"Required stat: {RequiredStat}\n\r" +
               $"Outcome:\n" +
               $"Positive: {PositiveOutcome}\n" +
               $"Negative: {NegativeOutcome}\n\r";
    }
}

File: Dsl/dnd/lib/dnd/action.rb
# frozen_string_literal: true
require_relative 'exceptions'
require_relative 'outcome'
require_relative 'stat'

class Action
  def initialize(&block)
    @description = nil
    @required_stat = nil
    @positive_outcome = nil
    @negative_outcome = nil
    @required_tag = nil
    instance_eval(&block)
  end

  def desc(description)
    @description = description
  end

  def require_to_have(stat_type, value)
    @required_stat = Stat.new(stat_type, value)
  end
  def required_tag(tag)
    @required_tag = tag
  end
  def in_case_of_success(&block)
    @positive_outcome = Outcome.new(&block)
  end

  def in_case_of_failure(&block)
    @negative_outcome = Outcome.new(&block)
  end

  def to_hash
    {
      description: @description,
      required_stat: @required_stat.nil? ? {} : @required_stat.to_hash,
      positive_outcome: @positive_outcome.nil? ? {} : @positive_outcome.to_hash,
      negative_outcome: @negative_outcome.nil? ? {} : @negative_outcome.to_hash,
    }
  end

  def validate
    not_nullables = [:description, :positive_outcome]
    not_nullables.each do |attr|
      raise Exceptions::NotNullableAttributeError.new("#{attr} cannot be nil") if instance_variable_get("@#{attr}").nil?
    end
    @required_stat.validate if @required_stat
    @positive_outcome.validate if @positive_outcome
    @negative_outcome.validate if @negative_outcome
  end
end
